### PR TITLE
feat(subscription): optional use of AzAPI for Subscription creation

### PR DESCRIPTION
**Pull request by danwatco**
# Overview/summary

This PR adds functionality to create a subscription using the `azapi_resource` resource type rather than using `azurerm_subscription`. This is useful in situations where the principal running Terraform doesn't have access to the default management group where subscriptions are created. By using the API, we can provide the `managementGroupId` parameter and the subscription is then created in the provided management group, avoiding the need to have creation permissions underneath the default management group.

Written unit and deployment tests (feedback welcome, my first time writing Go tests). Unable to run deployment tests so needs some testing there. 

## Testing evidence
image

image

There is some concerns that the management group association may cause an import error but yet to be tested.

## As part of this pull request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-lz-vending/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-lz-vending/issues), for tracking and closure.
- [x] Run and `make fmt` & `make docs` to format your code and update documentation.
- [x] Created unit and deployment tests and provided evidence.
- [x] Updated relevant and associated documentation.

### CODE REVIEW

**Comment by matt-FFFFFF** [comment]
Thanks @danwatco this looks great. Appreciate you taking the time to create some tests too.

I think one scenario that needs to be tested with deployment tests is one of using AzAPI to create the subscription alias and then AzureRM to manage the management group association.

If you are happy to copy one of the other deployment tests in the subscription submodule then feel free to modify that, if not I can do it.

**Comment by danwatco** [comment]
> Thanks @danwatco this looks great. Appreciate you taking the time to create some tests too.
> 
> I think one scenario that needs to be tested with deployment tests is one of using AzAPI to create the subscription alias and then AzureRM to manage the management group association.
> 
> If you are happy to copy one of the other deployment tests in the subscription submodule then feel free to modify that, if not I can do it.

Have had a shot here - feel free to change/update

File: modules/subscription/main.tf [hunk 1]
```diff
@@ -1,17 +1,42 @@
...
+        tags              = var.subscription_tags
+      }
+    }
+  })
+}
+
 # This resource ensures that we can manage the management group for the subscription
 # throughout its lifecycle.
 resource "azurerm_management_group_subscription_association" "this" {
   count               = var.subscription_management_group_association_enabled ? 1 : 0
```
Comment (jaredfholgate) [inline modules/subscription/main.tf:37]: @luke-taylor You don't need this block if using AzAPI, so you can count = 0 it.

**Review by jaredfholgate — CHANGES_REQUESTED** [review]

**Comment by matt-FFFFFF** [comment]
Thanks for this @danwatco - @luke-taylor has implemented this in #253

**Pull request closed without merging**